build_resolvers/lib/src/analysis_driver_model.dart

```diff
@@ -19,36 +25,148 @@ import 'package:build/build.dart';
+    if (assetId == null || !resourceProvider.getFile(assetId.asPath).exists) {
+      return null;
+    }
+
+    return assetId;
+  }
 
   /// Updates [resourceProvider] and the analysis driver given by
   /// `withDriverResource`  with updated versions of [entryPoints].
   ///
   /// If [transitive], then all the transitive imports from [entryPoints] are
   /// also updated.
+  ///
+  /// Notifies [buildStep] of all inputs that result from analysis. If
+  /// [transitive], this includes all transitive dependencies.
+  ///
+  /// If while finding transitive deps a `.transitive_deps` file is
+  /// encountered next to a source file then this cuts off the reporting
+  /// of deps to the [buildStep], but does not affect the reporting of
+  /// files to the analysis driver.
   Future<void> performResolve(
       BuildStep buildStep,
       List<AssetId> entryPoints,
       Future<void> Function(
               FutureOr<void> Function(AnalysisDriverForPackageBuild))
           withDriverResource,
-      {required bool transitive});
+      {required bool transitive}) async {
+    /// TODO(davidmorgan): add test coverage for whether transitive
+    /// sources are read when [transitive] is false, fix the implementation
+    /// here.
+    /// TODO(davidmorgan): add test coverage for whether
+    /// `.transitive_deps` files cut off the reporting of deps to the
+    /// [buildStep], fix the implementation here.
+
+    // Find transitive deps, this also informs [buildStep] of all inputs).
+    final ids = await _expandToTransitive(buildStep, entryPoints);
+
+    // Apply changes to in-memory filesystem.
+    for (final id in ids) {
+      if (await buildStep.canRead(id)) {
+        final content = await buildStep.readAsString(id);
+
+        /// TODO(davidmorgan): add test coverage for when a file is
+        /// modified rather than added, fix the implementation here.
+        resourceProvider.newFile(id.asPath, content);
+      } else {
+        if (resourceProvider.getFile(id.asPath).exists) {
+          resourceProvider.deleteFile(id.asPath);
+        }
+      }
+    }
+
+    // Notify the analyzer of changes.
+    await withDriverResource((driver) async {
+      for (final id in ids) {
+        // TODO(davidmorgan): add test coverage for over-notification of
+        // changes, fix the implementaion here.
+        driver.changeFile(id.asPath);
+      }
+      await driver.applyPendingFileChanges();
+    });
+  }
+
+  /// Walks the import graph from [ids], returns full transitive deps.s
+  Future<Set<AssetId>> _expandToTransitive(
+      AssetReader reader, Iterable<AssetId> ids) async {
+    final result = <AssetId>{};
+    await __expandToTransitive(reader, ids, result);
+    return result;
+  }
+
+  /// Walks the import graph from [ids], ignoring nodes already in [result].
+  ///
+  /// Call with [result] empty to add full transitive deps to [result].
+  Future<void> __expandToTransitive(
```

Comment:
Not a big fan of the double underscore naming.

Reply:
Oh. Apparently I thought this was going to be needed for a recursive call, but actually it's iterative. So no need for this method at all, merged into the one above. Thanks!